Allow setting partitioner per table

haaawk opened the pull request:
This PR makes it possible to enable the usage of different partitioner for each table. If no table-specific partitioner is set for a given table then a default partitioner is used.

The PR is composed of the following parts:
1. Introduction of schema::get_partitioner that still returns dht::global_partitioner
2. Replacement of all the usage of dht::global_partitioner with schema::get_partitioner
3. Making it possible to set table-specific partitioner in a schema_builder
4. Remove all the places that were setting default partitioner except for main.cc (mostly tests)
5. Move default partitioner from i_partitioner to schema.cc and hide it from the rest of the codebase
6. Remove dht::global_partitioner

After this PR there's no such thing as global partitioner at all. There is only a default partitioner but it still has to be accessed through schema::get_partitioner.

There are some intermediate states in which i_partitioner is stored as shared_ptr in the schema but the final version keeps it by const&.

The PR does not enable per table partitioner end-to-end. Just the internals of the single node are covered. I still have to deal with:
1. Making sure a table has the same partitioner on each node
2. Allowing user to set up a table-specific partitioner on table
3. Signal driver about what partitioner is used by a given table
4. Persist partitioner info for each table that does not use default partitioner.

Fixes #5493 

Tests: unit(dev, release, debug), dtest(byo)

haaawk commented:
Successful BYO run can be found here: https://jenkins.scylladb.com/job/scylla-master/job/byo/job/byo_build_tests_dtest/922/

haaawk commented:
@asias Could you please have a look at streaming and repair changes?

avikivity commented on schema.cc:
```diff
@@ -36,6 +36,7 @@
 #include "partition_slice_builder.hh"
 #include "database.hh"
```
Missing changelog.

haaawk replied:
Added commit message.

avikivity commented on schema.hh:
```diff
@@ -809,6 +815,8 @@ public:
     }
 
+    dht::i_partitioner& get_partitioner() const;
+
```
How does this interact with schema digests? Should we hash it into the digest (but not do that if the default partitioner is used, to avoid disturbing the current digests)?

If we every allow altering the partitioner, we must, but it seems unlikely that we ever can.

@tgrabiec opinion please

haaawk replied:
I will rename this PR. Current description is not clear enough. The PR does not enable per table partitioner end-to-end. Just the internals of the single node are covered. I still have to deal with:
1. Making sure a table has the same partitioner on each node
2. Allowing user to set up a table-specific partitioner on table
3. Signal driver about what partitioner is used by a given table
4. Persist partitioner info for each table that does not use default partitioner.

avikivity replied:
For (1), it depends on the use case. For example if cdc logs have their own partitioner, then all nodes will make the same decision when they create the log table and there is no need for extra communication.

(3) is very important, and also unfortunately quite difficult. At least everything still works even without it.

haaawk replied:
I agree. For CDC everything should just work because partitioner will be fixed. The driver will be a bit less efficient because token_aware_policy probably won't work but we will fix it incrementally.

tgrabiec replied:
@avikivity I have a weak preference for including the partitioner name in the digest, even if we don't allow changing it. Digest should be consistent with equality and schemas with different partitioners are not equal.

avikivity replied:
Yes. But how do we deal with preexisting digests? They don't include the partitioner name. If it was just the name, then we could special-case the murmur3 partitioner not to be included, but there are also parameters, which vary from table to table.

tgrabiec replied:
For mixed clusters, until the whole cluster is upgraded to the version with per-table partitioners we can ignore the field when digesting.

avikivity replied:
Won't there be a lot of log noise because not all nodes switch to the new hash at the same instant?

tgrabiec replied:
There could be schema pulls as a result, yes

avikivity replied:
ok. @haaawk please do so, and test upgrades under a light load to see that it works and that log spam is not crazy.

haaawk replied:
Will do. I'm just wondering why you hate me so much :D

nyh replied:
@haaawk was this done?

haaawk replied:
Yes. partitioner name is included into digest now but only once the whole cluster is migrated to the new version that supports per table partitioners.

avikivity commented on dht/sharder.hh:
```diff
@@ -0,0 +1,134 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
```
You can drop the Apache license here - all the code is new so is only covered by the AGPL.

haaawk replied:
done.

avikivity commented on dht/i_partitioner.cc:
```diff
@@ -290,8 +290,9 @@ ring_position_exponential_sharder::next(const schema& s) {
     return std::make_optional(std::move(ret));
 }
```
changelog

haaawk replied:
Next patch moves the sharders to separate header. I didn't want to move unused code but didn't want to silently remove it together with move so I split it into two commits. One removing unused code and second moving the rest to separate header. Will add the changelog.

denesb replied:
AFAIK this was used for range-scans, before that stateful range scan implementation, which doesn't use them anymore.

haaawk replied:
Thanks @denesb

avikivity commented on dht/sharder.hh:
```diff
@@ -116,8 +116,6 @@ class selective_token_range_sharder {
     dht::token _start_token;
     std::optional<range_bound<dht::token>> _start_boundary;
```
changelog

asias replied:
The patch changes repair/ but the subject says selective_token_range_sharder
```
commit 209f912f954bd632a330e3372e58a80a6e3bc083
Author: Piotr Jastrzebski <piotr@scylladb.com>
Date:   Thu Dec 19 19:08:30 2019 +0100

    selective_token_range_sharder: stop calling global_partitioner()
    
    Signed-off-by: Piotr Jastrzebski <piotr@scylladb.com>
---
 repair/repair.hh |  2 ++
 repair/repair.cc | 30 ++++++++++++++++++++++--------
 2 files changed, 24 insertions(+), 8 deletions(-)
```

haaawk replied:
I hate github and its way of showing comments on wrong commits!!!

haaawk replied:
I added a commit message to explain the changes in repair.

avikivity commented on repair/repair.cc:
```diff
@@ -76,14 +76,16 @@ std::ostream& operator<<(std::ostream& out, row_level_diff_detect_algorithm algo
 
-static std::vector<sstring> list_column_families(const database& db, const sstring& keyspace) {
+static std::pair<std::vector<sstring>, std::vector<schema_ptr>> list_column_families(const database& db, const sstring& keyspace) {
     std::vector<sstring> ret;
```
schema contains the name, so it's enough to return the schema vector.

haaawk replied:
You're obviously right. The reason I decided to return both is to minimize the change.

list_column_families is used in line 1415. In there you can see that code following the line works on a vector of column family names. There are two cases those names come from: from options (line 1401/1406) or from list_column_families (line 1415). To unify the two cases I decided to:
1. In the first option when cfs come from options, I generate vector of schema_ptrs from names
2. In the second option I generate both vector of names and vector of schema_ptrs in list_column_families
This way rest of the code can work on those two vectors no matter which case it is.
There is an alternative to return just a vector of schema_ptrs and change all the rest of the code to use it instead of the vector of names (cfs) but that is much bigger change and I wanted to minimize the friction.
@avikivity and @nyh do you want me to change all that code so that we can return single vector of schemas here?

haaawk replied:
With a check being moved to repair_info, change to list_column_families will be reverted so we can consider this commend addressed.

avikivity commented on repair/repair.cc:
```diff
@@ -1280,7 +1284,7 @@ static future<> do_repair_ranges(lw_shared_ptr<repair_info> ri) {
                 ri->ranges_index, ri->ranges.size(), ri->id, ri->shard, ri->keyspace, ri->cfs, range);
-            return do_with(dht::selective_token_range_sharder(range, ri->shard), [ri] (auto& sharder) {
+            return do_with(dht::selective_token_range_sharder(ri->partitioner, range, ri->shard), [ri] (auto& sharder) {
                 return repeat([ri, &sharder] () {
```
This function repairs multiple tables, so how can it accept a single partitioner?

Either it should use the partitioners from the tables, or it should verify that the table partitioners match the input partitioner.

avikivity replied:
Actually, it does verify!

avikivity commented on repair/repair.cc:
```diff
@@ -1397,29 +1401,39 @@ static int do_repair_start(seastar::sharded<database>& db, sstring keyspace,
+            throw std::runtime_error(
+                    format("All column families repaired together have to have the same partitioner. Different partitioners found: {} and {}",
+                        schemas[idx - 1]->get_partitioner(), schemas[idx]->get_partitioner()));
+        }
```
Better mention the table names too.

asias replied:
s/column families/ tables.

asias replied:
With this patch,  `nodetool repair keyspace` will fail if tables with different partitioner are inside this keyspace. In this case, user has to manually specify the tables and repair table by table. What if there are 1000 tables. 

I think this is a regression.

haaawk replied:
This is true @asias good thing is that by default all the tables use default partitioner so the problem does not exist. It's on the user to not create table with custom partitioner if they find it problematic to repair them by manually selecting tables. I don't think it's a regression. It does not affect any of the existing keyspaces. Only when user decides to use the new functionality of per table partitioner then they will face the problem but that's part of the price for the feature. No feature use - no cost. Zero cost abstraction :)

denesb replied:
Is the partitioner even something that can be set by the user (via CQL). I didn't see any code that would persist and load the partitioner into the system schema tables.

Also, supporting this would open a whole can of worms, as a schema later can now possibly change token ownership (even if only within a single node).

haaawk replied:
It is not possible at the moment. Initial user will be internal (CDC). We may later want to allow users to set per table partitioners but only during create table and disallow changing it using alter table.

haaawk replied:
I changed the column families to tables and added table names.

avikivity commented on repair/repair.hh:
```diff
@@ -183,6 +184,7 @@ public:
 public:
     repair_info(seastar::sharded<database>& db_,
+            const dht::i_partitioner& p,
             const sstring& keyspace_,
```
Please add a comment here that the partitioner must match the tables. Maybe it's better to move the check here, it's cleaner not to depend on the caller.

asias replied:
In https://github.com/scylladb/scylla/pull/4975, we use repair_info in sync_data_using_repair. If we move the check into repair_info, we can avoid duplicating the check code. So please do so.

haaawk replied:
I moved the check to repair_info constructor.

avikivity commented on dht/i_partitioner.cc:
```diff
@@ -248,10 +248,6 @@ ring_position_exponential_sharder::ring_position_exponential_sharder(const i_par
     }
 }
```
changelog

haaawk replied:
added.

avikivity commented on alternator/executor.cc:
```diff
@@ -1155,7 +1155,7 @@ std::optional<shard_id> rmw_operation::shard_for_execute(bool needs_read_before_
     // If we're still here, cas() *will* be called by execute(), so let's
```
changelog

haaawk replied:
added.

haaawk replied:
done.

avikivity commented on alternator/executor.cc:
```diff
@@ -1027,7 +1027,7 @@ static lw_shared_ptr<query::read_command> read_nothing_read_command(schema_ptr s
 }
 
```
changelog should explain why this is a good thing.

haaawk replied:
I added a commit message.

avikivity commented on sstables/sstables.cc:
```diff
@@ -490,17 +490,17 @@ typename disk_set_of_tagged_union<TagType, Members...>::serdes disk_set_of_tagge
 future<>
-parse(sstable_version_types v, random_access_reader& in, disk_set_of_tagged_union<TagType, Members...>& s) {
+parse(schema_ptr schema, sstable_version_types v, random_access_reader& in, disk_set_of_tagged_union<TagType, Members...>& s) {
     using disk_set = disk_set_of_tagged_union<TagType, Members...>;
```
You are copying schema_ptr:s like crazy. They're pretty lightweight but still that bloats the code with calls to the schema destructor that will never happen.

Can you capture the schema by reference? Need to make sure the top-level caller to parse holds a steay pointer to the schema.

haaawk replied:
I changed it to const& the only caller takes schema from sstable so it's safe. @nyh I hope you won't mind me not replying to all the move comments - I appreciate the time you took to find all of them but happily Avi's suggestion solves the problem.

avikivity commented on sstables/sstables.cc:
```diff
@@ -1043,7 +1043,7 @@ future<> sstable::read_simple(T& component, const io_priority_class& pc) {
             auto r = make_lw_shared<file_random_access_reader>(std::move(fi), size, sstable_buffer_size);
-            auto fut = parse(_version, *r, component);
+            auto fut = parse(_schema, _version, *r, component);
             return fut.finally([r] {
```
For example, this caller is safe since it picks up the schema from the sstable, and an sstable's schema doesn't change.

avikivity commented on repair/repair.cc:
```diff
@@ -1218,13 +1217,12 @@ struct repair_options {
+    void set_ranges(std::unordered_map<sstring, sstring>& options,
+            dht::i_partitioner& partitioner) {
+        auto it = options.find(RANGES_KEY);
         if (it == options.end()) {
```
The changelog says "repair: stop calling global_partitioner()" but that's not what I see here.

nyh replied:
Right. I don't understand this patch at all. Also, @asias please review this patch that touches repair in non-obvious (to me) ways.

asias replied:
I added some comments.

haaawk replied:
This change is gone once I moved the check to repair_info constructor.

avikivity commented on test/boost/mutation_test.cc:
```diff
@@ -2942,10 +2942,9 @@ void run_compaction_data_stream_split_test(const schema& schema, gc_clock::time_
 } // anonymous namespace
```
changelog: why should we stop taking partitioner in constructor?

haaawk replied:
added more explanation.

haaawk commented:
This PR made mutation_write_test flaky so I fixed that and re-run unit tests (dev, release, debug). The only changed file was test/boost/mutation_writer_test.cc

avikivity commented on service/storage_service.cc:
```diff
@@ -3608,7 +3608,7 @@ std::vector<gms::inet_address>
     sstables::key_view key_view = sstables::key_view(bytes_view(reinterpret_cast<const signed char*>(key.c_str()), key.size()));
-    dht::token token = dht::global_partitioner().get_token(key_view);
+    dht::token token = _db.local().find_schema(keyspace, key)->get_partitioner().get_token(key_view);
     return get_natural_endpoints(keyspace, token);
```
s/key/cf/? did this pass tests?

haaawk replied:
It did. Both unit at all level and BYO. Weird.

nyh replied:
I don't understand how... It seems to me like *any* use of get_natural_endpoints would have thrown on unfound table, no?

haaawk replied:
Maybe we don't have tests that cover this. The only usage of this method I found was in api::storage_service maybe it's never called in tests.

haaawk replied:
Fixed.

avikivity commented on test/boost/partitioner_test.cc:
```diff
@@ -327,8 +327,6 @@ static
 void
 test_exponential_sharder(const dht::i_partitioner& part, const schema& s, const dht::partition_range& pr) {
 
```
Changelog should explain why we can do that.

haaawk replied:
done.

avikivity commented on cdc/generation.cc:
```diff
@@ -26,9 +26,9 @@
 
 #include "keys.hh"
 #include "schema_builder.hh"
```
Why should cdc stop using partitioners?

haaawk replied:
Because it just needs sharding which does not depend on partitioner any more. I will add the explanation to the commit message.

avikivity commented on test/boost/sstable_resharding_test.cc:
```diff
@@ -30,8 +30,7 @@ static inline std::vector<std::pair<sstring, dht::token>> token_generation_for_s
-    dht::set_global_partitioner(to_sstring("org.apache.cassandra.dht.Murmur3Partitioner"));
-    dht::murmur3_partitioner partitioner(smp::count, 0);
+    dht::murmur3_partitioner partitioner(smp::count, 12);
 
```
This is more than "sstable_resharding_test: stop calling set_global_partitioner".

haaawk replied:
changed.

avikivity commented on schema.hh:
```diff
@@ -635,6 +636,7 @@ private:
         // It is set locally on a system tables that should be extra durable
         bool _wait_for_sync = false; // true if all writes using this schema have to be synced immediately by commitlog
+        std::shared_ptr<dht::i_partitioner> _partitioner;
     };
```
Why are you starting with std::shared_ptr? Not even seastar::shared_ptr?

haaawk replied:
It's just an intermediate step to break the cyclic dependency. It's replaced with const& in the following patches. I tried to explain it both on commit message and PR description but apparently it didn't work.

nyh replied:
Ok, but it's just strange it uses std::shared_ptr<> and <memory> instead of seastar::shared_ptr?

avikivity replied:
I'm good at complaining that changelogs are missing descriptions, but pretty bad at actually reading them.

Agree you should use a regular shared_ptr, it's just strange.

avikivity commented on schema.cc:
```diff
@@ -101,11 +102,21 @@ std::ostream& operator<<(std::ostream& os, ordinal_column_id id)
+    }
+    return *default_partitioner;
 }
 
```
It looks like you have two ways to set the default partitioner, one from set_default_partitioner() and one here. This means that schemas created before set_default_partitioner() is called may have the wrong partitioner.

The code here is racy. If multiple shards hit it at the same time, then each will try to write to the default_partitioner global.

haaawk replied:
This is removed in the following patch. But you're right. It used to be the same before though in global_partitioner. It's a fair point though. I need to make partitioners cache thread-local otherwise it's also racy.

avikivity commented on schema_builder.hh:
```diff
@@ -237,8 +237,9 @@ public:
+    schema_builder& with_partitioner(sstring name, unsigned shard_count, unsigned sharding_ignore_msb_bits) {
+        auto ptr = dht::make_partitioner(name, shard_count, sharding_ignore_msb_bits);
+        _raw._partitioner = std::shared_ptr<dht::i_partitioner>(ptr.release());
         return *this;
```
Won't this create lots of partitioner objects, even though they're all the same?

avikivity replied:
I don't understand why you use std::shared_ptr<> if you then call release.

haaawk replied:
See next patch. They will be cached.

avikivity replied:
Please drop the std::shared_ptr. I don't know how to review something that appears and disappears, because I can't tell if all the problems introduced by this patch are later resolved. I don't get a clear picture of the end result.

haaawk replied:
Do you want me to drop shard_ptr from return value of make_partitioner or from _raw._partitioner?

avikivity replied:
Both. The whole thing confuses me.

haaawk replied:
Sure. I can do that but it will require squashing at least 3 commits into one here. If you're ok with that - I'm happy to do it.

avikivity commented on schema.cc:
```diff
@@ -19,6 +19,7 @@
  * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
  */
```
changelog, what problem are you trying to solve and how are you going about it.

haaawk replied:
done.

nyh commented on sstables/sstables.cc:
```diff
@@ -320,28 +320,28 @@ inline void write(sstable_version_types v, file_writer& out, const utils::UUID&
-    return f.then([v, &in, &s, len = std::move(len)] {
-        return parse(v, in, *len, s.value);
+    auto f = parse(schema, v, in, *len);
+    return f.then([v, schema, &in, &s, len = std::move(len)] {
```
I think it is fine to use schema=std::move(schema) here, and std::move(schema) below for, again, a bit less reference counting?

nyh commented on sstables/sstables.cc:
```diff
@@ -320,28 +320,28 @@ inline void write(sstable_version_types v, file_writer& out, const utils::UUID&
     auto f = read_vint(in, *len);
-    return f.then([v, &in, &s, len = std::move(len)] {
-        return parse(v, in, *len, s.value);
+    return f.then([v, schema, &in, &s, len = std::move(len)] {
```
As usual, I think you can use schema=std::move(schema) here and std::move(schema) below.

nyh commented on sstables/sstables.cc:
```diff
@@ -320,28 +320,28 @@ inline void write(sstable_version_types v, file_writer& out, const utils::UUID&
     auto f = read_vint(in, *len);
-    return f.then([v, &in, &arr, len = std::move(len)] {
-        return parse(v, in, *len, arr.elements);
+    return f.then([v, s, &in, &arr, len = std::move(len)] {
```
again I think you can use s=std::move(s) and below std::move(s)

nyh commented on sstables/sstables.cc:
```diff
@@ -355,21 +355,21 @@ future<> parse(sstable_version_types v, random_access_reader& in, disk_array_vin
     auto eoarr = [count, len] { return *count == len; };
 
-    return do_until(eoarr, [v, count, &in, &arr] {
+    return do_until(eoarr, [v, s, count, &in, &arr] {
```
again s=std::move(s) and below std::move(s)?

nyh commented on sstables/sstables.cc:
```diff
@@ -389,27 +389,27 @@ parse(sstable_version_types, random_access_reader& in, Size& len, utils::chunked
-    auto f = parse(v, in, *len);
-    return f.then([v, &in, &arr, len] {
+    auto f = parse(s, v, in, *len);
+    return f.then([v, s, &in, &arr, len] {
```
I *think* it's find to do in this line s=std::move(s), and below std::move(s).

nyh commented on sstables/sstables.cc:
```diff
@@ -389,27 +389,27 @@ parse(sstable_version_types, random_access_reader& in, Size& len, utils::chunked
-future<> parse(sstable_version_types v, random_access_reader& in, Size& len, std::unordered_map<Key, Value>& map) {
-    return do_with(Size(), [v, &in, len, &map] (Size& count) {
+future<> parse(schema_ptr s, sstable_version_types v, random_access_reader& in, Size& len, std::unordered_map<Key, Value>& map) {
+    return do_with(Size(), [v, s, &in, len, &map] (Size& count) {
```
can std::move(s)

nyh commented on sstables/sstables.cc:
```diff
@@ -389,27 +389,27 @@ parse(sstable_version_types, random_access_reader& in, Size& len, utils::chunked
+    return do_with(Size(), [v, s, &in, len, &map] (Size& count) {
         auto eos = [len, &count] { return len == count++; };
-        return do_until(eos, [v, len, &in, &map] {
+        return do_until(eos, [v, s, len, &in, &map] {
```
can std::move(s)?

nyh commented on sstables/sstables.cc:
```diff
@@ -389,27 +389,27 @@ parse(sstable_version_types, random_access_reader& in, Size& len, utils::chunked
 
-            return do_with(kv(), [v, &in, &map] (auto& el) {
-                return parse(v, in, el.key, el.value).then([&el, &map] {
+            return do_with(kv(), [v, s, &in, &map] (auto& el) {
```
can std::move(s) here and in the line below?

nyh commented on sstables/sstables.cc:
```diff
@@ -418,16 +418,16 @@ future<> parse(sstable_version_types v, random_access_reader& in, Size& len, std
-future<> parse(sstable_version_types v, random_access_reader& in, std::pair<First, Second>& p) {
-    return parse(v, in, p.first, p.second);
+future<> parse(schema_ptr s, sstable_version_types v, random_access_reader& in, std::pair<First, Second>& p) {
+    return parse(s, v, in, p.first, p.second);
```
std::move(s)

nyh commented on sstables/sstables.cc:
```diff
@@ -418,16 +418,16 @@ future<> parse(sstable_version_types v, random_access_reader& in, Size& len, std
-    return f.then([v, &in, &h, w = std::move(w)] {
-        return parse(v, in, *w, h.map);
+    auto f = parse(s, v, in, *w);
+    return f.then([v, s, &in, &h, w = std::move(w)] {
```
std::move(s) here and in the line below?

nyh commented on sstables/sstables.cc:
```diff
@@ -446,9 +446,9 @@ template <typename DiskSetOfTaggedUnion, typename Member>
+    virtual future<> do_parse(schema_ptr s, sstable_version_types version, random_access_reader& in, value_type& v) const {
         v = Member();
-        return parse(version, in, boost::get<Member>(v).value);
+        return parse(s, version, in, boost::get<Member>(v).value);
```
std::move(s)

nyh commented on sstables/sstables.cc:
```diff
@@ -467,12 +467,12 @@ struct disk_set_of_tagged_union<TagType, Members...>::serdes {
             return in.read_exactly(size).discard_result();
         } else {
-            return i->second->do_parse(v, in, value).then([tag, &s, &value] () mutable {
+            return i->second->do_parse(schema, v, in, value).then([tag, &s, &value] () mutable {
```
std::move(schema)?

nyh commented on sstables/sstables.cc:
```diff
@@ -490,17 +490,17 @@ typename disk_set_of_tagged_union<TagType, Members...>::serdes disk_set_of_tagge
     using value_type = typename disk_set::value_type;
-    return do_with(0u, 0u, 0u, value_type{}, [&] (key_type& nr_elements, key_type& new_key, unsigned& new_size, value_type& new_value) {
-        return parse(v, in, nr_elements).then([&, v] {
+    return do_with(0u, 0u, 0u, value_type{}, [&, schema] (key_type& nr_elements, key_type& new_key, unsigned& new_size, value_type& new_value) {
```
please think on all these schema captures if we can do a move capture. There are too many of these in this function for me to keep track ;-)

nyh commented on sstables/sstables.cc:
```diff
@@ -550,20 +550,20 @@ future<> parse(sstable_version_types v, random_access_reader& in, summary& s) {
 
-            return do_with(int(0), [&in, &s] (int& idx) mutable {
-                return do_until([&s] { return s.entries.size() == s.header.size; }, [&s, &in, &idx] () mutable {
+            return do_with(int(0), [schema, &in, &s] (int& idx) mutable {
```
some of these schema captures, like this one, can probably be done with move. Not all of them, however :-(

nyh commented on sstables/sstables.cc:
```diff
@@ -619,42 +619,42 @@ future<summary_entry&> sstable::read_summary_entry(size_t i) {
-future<> parse(sstable_version_types v, random_access_reader& in, deletion_time& d) {
-    return parse(v, in, d.local_deletion_time, d.marked_for_delete_at);
+future<> parse(schema_ptr s, sstable_version_types v, random_access_reader& in, deletion_time& d) {
+    return parse(s, v, in, d.local_deletion_time, d.marked_for_delete_at);
```
std::move(s)

nyh commented on sstables/sstables.cc:
```diff
@@ -619,42 +619,42 @@ future<summary_entry&> sstable::read_summary_entry(size_t i) {
+future<> parse(schema_ptr s, sstable_version_types v, random_access_reader& in, std::unique_ptr<metadata>& p) {
     p.reset(new Child);
-    return parse(v, in, *static_cast<Child *>(p.get()));
+    return parse(s, v, in, *static_cast<Child *>(p.get()));
```
std::move(s)

nyh commented on sstables/sstables.cc:
```diff
@@ -672,10 +672,10 @@ inline void write(sstable_version_types v, file_writer& out, const statistics& s
     auto len = std::make_unique<uint32_t>();
 
-    auto f = parse(v, in, *len);
+    auto f = parse(s, v, in, *len);
```
std::move(s)

nyh commented on sstables/sstables.cc:
```diff
@@ -748,10 +748,10 @@ struct streaming_histogram_element {
     auto a = std::make_unique<disk_array<uint32_t, streaming_histogram_element>>();
 
-    auto f = parse(v, in, sh.max_bin_size, *a);
+    auto f = parse(s, v, in, sh.max_bin_size, *a);
```
std::move(s)

nyh commented on sstables/sstables.cc:
```diff
@@ -789,9 +789,9 @@ void write(sstable_version_types v, file_writer& out, const utils::streaming_his
-        return parse(v, in, ci.end);
+future<> parse(schema_ptr s, sstable_version_types v, random_access_reader& in, commitlog_interval& ci) {
+    return parse(s, v, in, ci.start).then([&ci, v, s, &in] {
+        return parse(s, v, in, ci.end);
```
std::move(s) here (unfortunately I don't know what we can do about the two copies in the line above).

nyh commented on sstables/sstables.cc:
```diff
@@ -800,16 +800,16 @@ void write(sstable_version_types v, file_writer& out, const commitlog_interval&
-      return do_with(uint32_t(), c.offsets.get_writer(), [v, &in, &c] (uint32_t& len, compression::segmented_offsets::writer& offsets) {
-        return parse(v, in, len).then([&in, &c, &len, &offsets] {
+      return do_with(uint32_t(), c.offsets.get_writer(), [v, s, &in, &c] (uint32_t& len, compression::segmented_offsets::writer& offsets) {
+        return parse(s, v, in, len).then([&in, &c, &len, &offsets] {
```
std::move(s)

nyh commented on sstables/index_entry.hh:
```diff
@@ -541,6 +542,7 @@ public:
 
 class index_entry {
 private:
+    std::reference_wrapper<const schema> _s;
```
Why is this a "reference_wrapper" and not a schema_ptr? Is it because (?) nothing in this code ever gets deferred? Can we be sure of that?

denesb replied:
I guess its needed because the class has a move assignment operator, which is incompatible with const members. We could just use a pointer instead (which is what `std::reference_wrapper<>` really is).

haaawk replied:
_s in index reader is const& so I didn't think of using schema_ptr (isn't it less efficient?). Botond is right that I had to use reference_wrapper because of the assignment.

nyh commented on streaming/stream_session.cc:
```diff
@@ -214,7 +214,7 @@ void stream_session::init_messaging_service_handler() {
                     };
                     //FIXME: discarded future.
-                    (void)mutation_writer::distribute_reader_and_consume_on_shards(s, dht::global_partitioner(),
+                    (void)mutation_writer::distribute_reader_and_consume_on_shards(s, s->get_partitioner(),
```
Again, a function which takes both a schema and a partitioner no longer needs to and can take just a schema. But I guess we can clean this up later.

haaawk replied:
Thanks - fixed.

nyh commented on test/lib/random_schema.cc:
```diff
@@ -981,7 +980,7 @@ std::vector<data_model::mutation_description::key> random_schema::make_pkeys(siz
     uint32_t i{0};
     while (keys.size() < n) {
-        keys.emplace(_partitioner.decorate_key(*_schema, partition_key::from_exploded(make_partition_key(i, val_gen))));
+        keys.emplace(_schema->get_partitioner().decorate_key(*_schema, partition_key::from_exploded(make_partition_key(i, val_gen))));
```
nitpick: I think at this point you already have dht::decorate_key() which does this a bit shorter.

haaawk replied:
Thanks. Fixed.

nyh commented on test/boost/sstable_datafile_test.cc:
```diff
@@ -1719,11 +1719,12 @@ static std::vector<std::pair<sstring, dht::token>> token_generation_for_shard(un
 
     key_and_token_pair.reserve(tokens_to_generate);
     dht::default_partitioner = std::make_unique<dht::murmur3_partitioner>(smp_count, ignore_msb);
+    dht::murmur3_partitioner partitioner(smp_count, ignore_msb);
```
How is this "partitioner" different from the default_partitioner defined right above it? Why do you need both?

haaawk replied:
dht::default_partitioner = ... is a leak in abstraction and it's removed in the following patch. I had to fix the test first to be able to hide default_partitioner and fix the abstraction leak.

nyh commented on test/boost/sstable_resharding_test.cc:
```diff
@@ -30,11 +31,12 @@ static inline std::vector<std::pair<sstring, dht::token>> token_generation_for_s
 
     key_and_token_pair.reserve(tokens_to_generate);
     dht::set_global_partitioner(to_sstring("org.apache.cassandra.dht.Murmur3Partitioner"));
+    dht::murmur3_partitioner partitioner(smp::count, 0);
```
If we have dht::set_global_partitioner(), is there also a dht::get_global_partitioner() so you don't need to repeat the same partitioner definition you defined right above it?

haaawk replied:
The goal of this PR is to remove global partitioner. All the commits are steps towards that direction. Here I change the test to create a partitioner it needs and use them directly instead of setting the global partitioner and use it implicitly.

nyh commented on test/boost/sstable_datafile_test.cc:
```diff
@@ -1768,9 +1768,10 @@ static shared_sstable sstable_for_overlapping_test(test_env& env, const schema_p
-static bool key_range_overlaps(sstring a, sstring b, sstring c, sstring d) {
-    auto range1 = create_token_range_from_keys(a, b);
-    auto range2 = create_token_range_from_keys(c, d);
+static bool key_range_overlaps(column_family_for_tests& cf, sstring a, sstring b, sstring c, sstring d) {
```
nitpick (feel free to ignore) - you didn't need cf here, schema& s would have been enough.

haaawk replied:
I agree. A partitioner would be even more specific but I just wanted to reduce the size of the change. If I take schema I would have to place cf->schema() in all the places the method is called and there's quite a few of them. I didn't want to repeat myself.

nyh commented on test/boost/mutation_writer_test.cc:
```diff
@@ -60,25 +60,26 @@ SEASTAR_TEST_CASE(test_multishard_writer) {
+                    partitioner,
                     std::move(source_reader),
-                    [&shards_after, error] (flat_mutation_reader reader) mutable {
+                    [&partitioner, &shards_after, error] (flat_mutation_reader reader) mutable {
```
How can we pass a schema_ptr by reference? Is it because we assume we run in a seastar::thread and "partitioner" above holds it?

haaawk replied:
We're taking partitioner by refererence - not schema_ptr. BTW this change goes away because schema is enough as you correctly mentioned on one of the previous commits.

nyh commented on test/boost/sstable_resharding_test.cc:
```diff
@@ -30,8 +30,7 @@ static inline std::vector<std::pair<sstring, dht::token>> token_generation_for_s
     key_and_token_pair.reserve(tokens_to_generate);
-    dht::set_global_partitioner(to_sstring("org.apache.cassandra.dht.Murmur3Partitioner"));
-    dht::murmur3_partitioner partitioner(smp::count, 0);
+    dht::murmur3_partitioner partitioner(smp::count, 12);
```
Oh, now you remove what I asked about. I guess you can conflate this patch with the other one I asked about.
But - why did you change from 0 in the previous patch to 12 here?

haaawk replied:
I merged this patch into the previous one you commented on.

nyh commented on schema.cc:
```diff
@@ -102,21 +102,30 @@ std::ostream& operator<<(std::ostream& os, ordinal_column_id id)
 
-std::unique_ptr<dht::i_partitioner> default_partitioner;
+std::map<std::tuple<sstring, unsigned, unsigned>, std::unique_ptr<dht::i_partitioner>> cached_partitioners;
+sstring default_partitioner_name = "org.apache.cassandra.dht.Murmur3Partitioner";
```
The *previous* patch claimed it was bout moving the default partitioner to schema.cc - but now I see it in this patch :-)

haaawk replied:
This is the next step. Previous patch moved default partitioner from i_partitioner to schema and this patch changes the way it's stored. It used to be stored as i_partitioner object but now it will be just a name and ignore_msb value. The i_partition object itself will be cached in cached_partitioners.

asias commented on repair/repair.cc:
```diff
@@ -1397,29 +1401,39 @@ static int do_repair_start(seastar::sharded<database>& db, sstring keyspace,
-        cfs = list_column_families(db.local(), keyspace);
+        std::tie(cfs, schemas) = list_column_families(db.local(), keyspace);
+    }
+    if (cfs.empty()) return id;
```
Use {} with if. 

The return here will skip the tracker::run() below which prints repair id has finished successfully, so

```
if (cfs.empty()) {
     rlogger.info("repair id {} completed successfully", id);
     return id;
}

or

if (cfs.empty()) {
     rlogger.info("repair id {} completed successfully: no tables to repair", id);
     return id;
}
```

haaawk replied:
Good idea. Thanks @asias! Done!

asias commented on repair/repair.cc:
```diff
@@ -1346,6 +1345,32 @@ static int do_repair_start(seastar::sharded<database>& db, sstring keyspace,
+    }
+    dht::i_partitioner& partitioner = schemas[0]->get_partitioner();
+
+    options.set_ranges(options_map, partitioner);
```
The partitioner is not used inside set_ranges. Why do you pass it?

haaawk replied:
This change is gone once I moved the check to repair_info constructor.

asias commented on repair/repair.cc:
```diff
@@ -1400,29 +1425,6 @@ static int do_repair_start(seastar::sharded<database>& db, sstring keyspace,
-                        schemas[idx - 1]->get_partitioner(), schemas[idx]->get_partitioner()));
-        }
-    }
-    dht::i_partitioner& partitioner = schemas[0]->get_partitioner();
```
Why do you want to move the code up? Even if there is a reason to do so, please split it out to a separate move-only patch. It is much easier to review.

haaawk replied:
This change is gone once I moved the check to repair_info constructor.

asias commented on repair/repair.cc:
```diff
@@ -1218,13 +1217,12 @@ struct repair_options {
-            std::unordered_map<sstring, sstring>& options,
-                        const sstring& key) {
-        auto it = options.find(key);
+    void set_ranges(std::unordered_map<sstring, sstring>& options,
```
why do you want to rename? ranges_opt, bool_opt, list_opt are parsing the options. The existing names are consistent. More importantly, this patch does not do what the subject say. Let's not add cosmetic changes and make the changes to other subsystems minimal.

haaawk replied:
This change is gone once I moved the check to repair_info constructor.

asias commented:
@haaawk I looked at the streaming and repair and storage_service/* part. Looks good in general. I left some comments in the individual patch.  Thanks!

git log --stat -p --ignore-all-space  origin/next.. -- repair/ streaming/  service/storage_service.*

haaawk commented:
Thank you @avikivity, @nyh and @asias. I will prepare a v2 with the fixes.

haaawk commented on dht/i_partitioner.cc:
```diff
@@ -290,8 +290,9 @@ ring_position_exponential_sharder::next(const schema& s) {
 }
 
-ring_position_range_vector_sharder::ring_position_range_vector_sharder(dht::partition_range_vector ranges)
+ring_position_range_vector_sharder::ring_position_range_vector_sharder(const dht::i_partitioner& p, dht::partition_range_vector ranges)
```
sharder depends on partitioner here: https://github.com/scylladb/scylla/pull/5750/commits/8d98b7bcd37219b9e95e196fedce86d1933cdae6#diff-11e7bcd8a00869edd8fa7b3a87ef38ccR127
Will add changelog.

haaawk commented on dht/i_partitioner.hh:
```diff
@@ -234,6 +234,8 @@ public:
     bool operator!=(const i_partitioner& o) const {
         return !(*this == o);
     }
+    friend std::ostream& operator<<(std::ostream&, const i_partitioner&);
```
removed.

denesb commented on cql3/statements/select_statement.cc:
```diff
@@ -860,8 +860,7 @@ ::shared_ptr<const service::pager::paging_state> indexed_table_select_statement:
     } else {
-        dht::i_partitioner& partitioner = dht::global_partitioner();
-        token_bytes = partitioner.get_token(*_schema, last_base_pk).data();
+        token_bytes = dht::get_token(*_schema, last_base_pk).data();
```
Could be folded with the previous patch (not a must).

denesb commented on db/size_estimates_virtual_reader.cc:
```diff
@@ -186,12 +186,12 @@ static system_keyspace::range_estimates estimate(const column_family& cf, const
 }
 
-future<std::vector<token_range>> get_local_ranges() {
+future<std::vector<token_range>> get_local_ranges(schema_ptr s) {
```
Virtual readers such as this usually have to take care of filtering out data that doesn't belong to the current shard according to the partitioner. Not doing so will violate ordering assumptions in the multishard reader. This filtering now requires the schema but I can't find the code that actually does it here.

haaawk replied:
This is patch was probably needed before I fixed the token representation on int64_t. Will remove this patch. Thanks.

denesb commented on multishard_mutation_query.cc:
```diff
@@ -592,7 +592,8 @@ static future<reconcilable_result> do_query_mutations(
                     mutation_reader::forwarding fwd_mr) {
-                return make_multishard_combining_reader(ctx, dht::global_partitioner(), std::move(s), pr, ps, pc, std::move(trace_state), fwd_mr);
+                    auto& partitioner = s->get_partitioner();
+                return make_multishard_combining_reader(ctx, partitioner, std::move(s), pr, ps, pc, std::move(trace_state), fwd_mr);
```
Yes, also having two arguments enables bugs related to passing a schema and a mismatching partitioner.

haaawk replied:
I agree with you both. Fixed. Thank you for catching. Unfortunately it can't be fixed before the ability to add partitioner to the schema is added because some tests use this method with dummy_partitioner. Will fix as a commit on top of the PR

denesb commented on repair/row_level.cc:
```diff
@@ -496,7 +496,7 @@ class repair_writer {
         auto& db = service::get_local_storage_service().db();
         table& t = db.local().find_column_family(_schema->id());
-        _writer_done[node_idx] = mutation_writer::distribute_reader_and_consume_on_shards(_schema, dht::global_partitioner(),
+        _writer_done[node_idx] = mutation_writer::distribute_reader_and_consume_on_shards(_schema, _schema->get_partitioner(),
```
Again, schema + partitioner passed to a method.

haaawk replied:
fixed.

denesb commented on schema.cc:
```diff
@@ -1461,7 +1461,7 @@ bytes token_column_computation::serialize() const {
 bytes_opt token_column_computation::compute_value(const schema& schema, const partition_key& key, const clustering_row& row) const {
-    dht::i_partitioner& partitioner = dht::global_partitioner();
+    dht::i_partitioner& partitioner = schema.get_partitioner();
     return partitioner.get_token(schema, key).data();
```
Don't we have `dht::get_token()` now?

haaawk replied:
fixed.

denesb commented on schema.cc:
```diff
@@ -101,11 +102,21 @@ std::ostream& operator<<(std::ostream& os, ordinal_column_id id)
+    if (!default_partitioner) {
+        default_partitioner = std::make_unique<dht::murmur3_partitioner>(smp::count, 12);
+    }
+    return *default_partitioner;
```
Why is all this done in the getter? Can't we do this in the constructor and make sure `_raw._partitioner` is never null, and points to `default_partitioner` instead?

haaawk replied:
This patch just moves code from one place to the other. Following patches change this.

denesb commented on schema.cc:
```diff
@@ -102,21 +102,30 @@ std::ostream& operator<<(std::ostream& os, ordinal_column_id id)
 }
 
-std::unique_ptr<dht::i_partitioner> default_partitioner;
+std::map<std::tuple<sstring, unsigned, unsigned>, std::unique_ptr<dht::i_partitioner>> cached_partitioners;
```
Missing a comment about why it is safe for this to *not* be `thread_local` and how is access orchestrated.

haaawk replied:
It is actually not safe to not be thread_local and I'm changing this. Thanks @denesb and @avikivity for pointing it out.

haaawk commented on dht/i_partitioner.cc:
```diff
@@ -248,10 +248,6 @@ ring_position_exponential_sharder::ring_position_exponential_sharder(const i_par
     }
 }
 
-ring_position_exponential_sharder::ring_position_exponential_sharder(partition_range pr)
```
It was unused. I think logically the patch fits here. Separating it now does not change much I guess.

haaawk commented on dht/i_partitioner.cc:
```diff
@@ -349,7 +349,7 @@ split_range_to_single_shard(const i_partitioner& partitioner, const schema& s, c
 future<utils::chunked_vector<partition_range>>
 split_range_to_single_shard(const schema& s, const partition_range& pr, shard_id shard) {
-    return split_range_to_single_shard(global_partitioner(), s, pr, shard);
+    return split_range_to_single_shard(s.get_partitioner(), s, pr, shard);
```
It's a fair point but we have a chicken and egg problem here. There are tests that use split_range_to_single_shard. At this point of the PR ability to set partitioner per schema is not yet added. I will fix this with a patch on top of last commit of this PR. But thanks for catching.

haaawk commented on service/storage_proxy.cc:
```diff
@@ -4827,7 +4827,7 @@ void storage_proxy::init_messaging_service() {
                                                               proposal = std::move(proposal), timeout] (schema_ptr schema) mutable {
             dht::token token = proposal.update.decorated_key(*schema).token();
-            unsigned shard = _db.local().shard_of(token);
+            unsigned shard = _db.local().shard_of(*schema, token);
```
It's a legacy. Notice that dht::shard_of was always a free function and database::shard_of was just a weird alias. I agree with your observation and will remove database::shard_of and replace all it's usages with dht::shard_of. Thanks.

haaawk commented on database.cc:
```diff
@@ -654,7 +654,7 @@ database::shard_of(const frozen_mutation& m) {
     // with token.
     schema_ptr schema = find_schema(m.column_family_id());
-    return shard_of(dht::global_partitioner().get_token(*schema, m.key(*schema)));
+    return shard_of(*schema, dht::global_partitioner().get_token(*schema, m.key(*schema)));
```
This commit only fixes shard_of calls. the commit that fixes get_token calls comes next.

haaawk commented on db/size_estimates_virtual_reader.cc:
```diff
@@ -141,7 +141,7 @@ static std::vector<sstring> get_keyspaces(const schema& s, const database& db, d
         range.slice(keyspaces, std::move(cmp)) | boost::adaptors::filtered([&s] (const auto& ks) {
             // If this is a range query, results are divided between shards by the partition key (keyspace_name).
-            return shard_of(dht::global_partitioner().get_token(s,
+            return shard_of(s, dht::global_partitioner().get_token(s,
```
see explanation above.

haaawk commented:
I rebased on top of master and addressed all the comments. Please re-review @avikivity, @nyh and @denesb

haaawk commented:
unit tests in dev and release are passing. I'm running debug and byo.

haaawk commented:
BYO has passed here: https://jenkins.scylladb.com/job/scylla-master/job/byo/job/byo_build_tests_dtest/923/

haaawk commented:
I had to fix mutation_reader_test.cc for debug unit tests to pass. Now all dev, release and debug pass. @avikivity @nyh @denesb please have a look

denesb commented on mutation_reader.cc:
```diff
@@ -1696,14 +1696,13 @@ reader_lifecycle_policy::try_resume(reader_concurrency_semaphore::inactive_read_
         tracing::trace_state_ptr trace_state,
         mutation_reader::forwarding fwd_mr) {
-    return make_flat_mutation_reader<multishard_combining_reader>(std::move(lifecycle_policy), partitioner, std::move(schema), pr, ps, pc,
+    return make_flat_mutation_reader<multishard_combining_reader>(std::move(lifecycle_policy), schema->get_partitioner(), std::move(schema), pr, ps, pc,
```
I think we should go all the way and remove `multishard_combining_reader::_partitioner` too.

haaawk replied:
done.

haaawk commented:
Rebased on top of master and removed multishard_combining_reader::_partitioner on @denesb request.

haaawk commented:
dev and release unit tests are passing.

haaawk commented:
debug unit tests all passed as well.

haaawk commented:
ping @avikivity @nyh

avikivity commented:
re-reviewing

avikivity commented on dht/sharder.hh:
```diff
@@ -0,0 +1,116 @@
+/*
+ * Modified by ScyllaDB
```
You can drop this line, it's only a requirement of the Apache License when importing code. All of the code here is ours.

haaawk replied:
done.

avikivity commented on sstables/compaction.cc:
```diff
@@ -744,8 +744,8 @@ class regular_compaction : public compaction {
     virtual flat_mutation_reader::filter make_partition_filter() const override {
-        return [] (const dht::decorated_key& dk){
-            return dht::shard_of(dk.token()) == engine().cpu_id();
+        return [s = _schema] (const dht::decorated_key& dk){
```
`&s = *_schema` eliminates some reference-count bookkeeping. The reader should hole _schema for us.

haaawk replied:
Is it guaranteed that a flat_mutation_reader will live longer than this lambda?

denesb replied:
The `_schema` member is from the `compaction` class, and that will outlive the reader.

haaawk replied:
nvm. It's schema of sstable.

haaawk replied:
done.

avikivity commented on table.cc:
```diff
@@ -592,8 +587,10 @@ flat_mutation_reader make_local_shard_sstable_reader(schema_ptr s,
         if (sst->is_shared()) {
-            using sig = bool (&)(const dht::decorated_key&);
-            reader = make_filtering_reader(std::move(reader), sig(belongs_to_current_shard));
+            auto filter = [s](const dht::decorated_key& dk) -> bool {
```
`&s = *s` to eliminate bookkeeping

haaawk replied:
done.

avikivity commented on schema.hh:
```diff
@@ -635,6 +636,7 @@ private:
         // The flag is not stored in the schema mutation and does not affects schema digest.
         // It is set locally on a system tables that should be extra durable
         bool _wait_for_sync = false; // true if all writes using this schema have to be synced immediately by commitlog
+        std::shared_ptr<dht::i_partitioner> _partitioner;
```
The comment above shoudl be extended to _partitioner. Digests and what they include or not are a major bug source, so we should be careful.

haaawk replied:
done.

avikivity commented on schema.cc:
```diff
@@ -102,21 +102,30 @@ std::ostream& operator<<(std::ostream& os, ordinal_column_id id)
+static const dht::i_partitioner& get_partitioner(const sstring& name, unsigned shard_count, unsigned ignore_msb) {
+    auto it = cached_partitioners.find({name, shard_count, ignore_msb});
+    if (it == cached_partitioners.end()) {
+        auto p = dht::make_partitioner(name, shard_count, ignore_msb);
```
This isn't really a cache, but interning. Items from the cache are never invalidated and reclaimed. 

That's not a problem since we have O(1) partitioners.

How does the partitioner interact with frozen_schema? I missed where we freeze and re-create it.

haaawk replied:
I renamed the variable to partitioners.

frozen_schema works in terms of schema_mutation which at the moment ignores partitioner. This is fine for now because the only custom partitioner will be for CDC and it will be always set to the same partitioner. Later when we will be making it possible to set any partitioner on any table we will have to include partitioner field in both schema_mutation and frozen_schema.

avikivity replied:
Not sure this is correct. global_schema_ptr uses frozen_schema, and in turn is used when queries or writes cross shards. So if someone queries from the cdc table from the wrong shard, we lose the partitioner.

haaawk replied:
You're right. It seems frozen_schema will have to handle partitioner as well.

nyh replied:
What's the conclusion of this discussion thread? Did you fix it (in this case please resolve the issue)? Left a FIXME and/or a issue? Or what?

haaawk replied:
This is fixed in this PR.

avikivity commented:
I recommend you split this into two, a large pull request with all the preparation where there isn't anything under contention (I made a few comments, but they are more or less cosmetic). The patches with the cache and std::shared_ptr are more difficult, so let's get the first one on and work on the second part afterwards.

haaawk commented:
Here's a first part split into a separate PR:
https://github.com/scylladb/scylla/pull/5822
It's rebased on top of the master and has your most recent comments fixed @avikivity

haaawk commented:
@avikivity I addressed your comments and rebased on top of master. I also removed the change that removed 'if' from schema::partitioner. It was problematic because schema created before default partitioner is set could use wrong partitioner. I will investigate more before resending that patch to be sure it's safe.

haaawk commented:
I added back the optimization that removes if from get_partitioner. I checked every usage of schema_builder and they all are safe.

haaawk commented:
All unit tests are passing (dev, release, debug).

haaawk commented:
ping @avikivity

avikivity commented on schema.hh:
```diff
@@ -635,6 +636,9 @@ private:
         bool _wait_for_sync = false; // true if all writes using this schema have to be synced immediately by commitlog
+        // Partitioner is not stored in the schema mutation and does not affect
+        // schema digest. It is also not set locally on a schema tables.
+        std::optional<std::reference_wrapper<const dht::i_partitioner>> _partitioner;
```
The std::optional later goes away?

haaawk replied:
Yes.

avikivity commented on test/boost/sstable_datafile_test.cc:
```diff
@@ -4175,39 +4174,36 @@ SEASTAR_TEST_CASE(sstable_owner_shards) {
                 | boost::adaptors::transformed([&] (auto shard) { return mut(shard); }));
-            dht::default_partitioner = std::make_unique<dht::murmur3_partitioner>(1, ignore_msb);
+            auto sst_gen = [&env, s, &tmp, gen, ignore_msb] () mutable {
+                auto schema = schema_builder(s).with_partitioner("org.apache.cassandra.dht.Murmur3Partitioner", 1, ignore_msb).build();
```
why is "1" correct here?

haaawk replied:
That's what it was before. See removed "dht::default_partitioner = std::make_unique<dht::murmur3_partitioner>(1, ignore_msb);"

haaawk replied:
Here's what used to be:
We used to:
1. set default_partitioner to dht::murmur3_partitioner(1, ignore_msb)
2. perform all the code below
3. clean up default partitioner to dht::murmur3_partitioner(smp_count, ignore_msb)

We now:
1. Create a new schema with dht::murmur3_partitioner(1, ignore_msb)
2. We perform all the code below with that schema which has the same effect as setting default_partitioner before. We switched all the places that were using the default partitioner to use schema now.
3. We don't have to cleanup default partitioner because we didn't change it.

nyh replied:
@avikivity are you satisfied by this response? It seems to me he is right that he didn't change the previous behavior, but I don't know if you're trying to understand why it was like this in the first place...

avikivity replied:
Agree that not changing behavior is correct here.

avikivity commented on test/boost/sstable_datafile_test.cc:
```diff
@@ -4175,39 +4174,36 @@ SEASTAR_TEST_CASE(sstable_owner_shards) {
         auto assert_sstable_owners = [&] (std::unordered_set<unsigned> expected_owners, unsigned ignore_msb, unsigned smp_count) {
             assert(expected_owners.size() <= smp_count);
             auto sst = make_shared_sstable(expected_owners, ignore_msb, smp_count);
-            dht::default_partitioner = std::make_unique<dht::murmur3_partitioner>(smp_count, ignore_msb);
```
I see here we used to use smp::count

haaawk replied:
I will answer on the previous comment.

avikivity commented on schema.hh:
```diff
@@ -638,7 +638,7 @@ private:
         // Partitioner is not stored in the schema mutation and does not affect
         // schema digest. It is also not set locally on a schema tables.
-        std::optional<std::reference_wrapper<const dht::i_partitioner>> _partitioner;
+        std::reference_wrapper<const dht::i_partitioner> _partitioner;
```
Good.

haaawk commented:
@avikivity I persisted partitioner for a table in scylla_tables. This means that it will be included in digest as @tgrabiec has requested. It also should work with global_schema_ptr now. Please have a look.

avikivity commented:
Did you verify that an upgrade recovers? run a cluster under a light load, upgrade its nodes, after the last upgrade there should be some noise from schema mismatches and then it settles.

haaawk commented:
No. Not yet. I'm implementing CDCPartitioner to have something to actually test this with. When all tables use default partitioner the new column is always empty.

haaawk commented:
I uploaded the code to give you more time to review it and leave some comments if you still have some.

haaawk commented:
Rebased on top of master.

haaawk commented:
@avikivity I've tested rolling upgrade with light load and everything went smooth. Do you have any other comments or can we merge this?

haaawk commented:
@avikivity ping

haaawk commented:
@avikivity ping pong

avikivity commented:
I'll give it a re-review.

avikivity commented on schema.cc:
```diff
@@ -840,6 +855,11 @@ bool thrift_schema::is_dynamic() const {
 
+schema_builder& schema_builder::with_partitioner(sstring name, unsigned shard_count, unsigned sharding_ignore_msb_bits) {
+    _raw._partitioner = get_partitioner(name, shard_count, sharding_ignore_msb_bits);
+    return *this;
```
What does it mean to associate a schema with shard_count and sharding_ignore_msb_bits? These are per-node settings. They aren't shared between different nodes for the same table.

avikivity replied:
Before we hashed the partitioner into the schema, it wasn't a problem, but now it is. Unless we only hash the partitioner name.

haaawk replied:
We only hash partitioner name. I agree that shard_count and sharding_ignore_msb_bits are per node. In this PR it's in schema because partitioner contains it. I'm preparing another PR on top of this one to separate partitioners and sharding logic for good.

avikivity replied:
Then everything's fine.

haaawk commented:
Rebased on top of master and resolved one conflict.

avikivity commented on db/schema_tables.cc:
```diff
@@ -305,23 +305,31 @@ schema_ptr tables() {
 
 // Holds Scylla-specific table metadata.
```
The changelog shoud explain about what this code does. I happen to know it will, but I'm not the only one who will try to read this commit one day and wonder about it.

avikivity replied:
Looks like the changelog thinks the column-censoring stuff is trivial.

avikivity commented on db/schema_tables.cc:
```diff
@@ -104,6 +104,7 @@ namespace db {
 
 schema_ctxt::schema_ctxt(const db::config& cfg)
```
I'm not reviewing any commits that don't have explanations any more.

haaawk replied:
I forgot - sorry about that.

avikivity replied:
The explanation repeats the subject line and adds no information.

haaawk replied:
What else would you like to see here? I've already explained why in 2-3 previous patches. Do you want me to repeat the same explanation in every commit?

avikivity replied:
Of course mindless repetition isn't helpful. But this patch does more than just add a field to the schema, it also handles rolling upgrades in a non-trivial way. Suppose that breaks and gets bisected to this patch, the lucky debugger will get zero help from the changelog about how this is supposed to work.

haaawk replied:
I gave it another try. Hopefully, now it's more comprehensive. @avikivity what do you think?

haaawk commented:
I added commit messages to every commit but "tests: remove unused murmur3 includes" @avikivity

haaawk commented:
Rebased on top of master.

haaawk commented:
Fixed grammar in commit message.

avikivity commented:
@haaawk github says that @nyh requested changes on this, are they all resolved?

haaawk commented:
I believe so @avikivity - the PR has changed a lot since I first uploaded it.

haaawk commented:
@avikivity @nyh do you want me to do anything else for this PR?

haaawk commented:
@avikivity @nyh please - this is blocking further work on CDC partitioner

haaawk commented:
Rebased on top of master and resolved conflicts.

nyh commented on db/schema_tables.cc:
```diff
@@ -305,23 +305,31 @@ schema_ptr tables() {
 // Holds Scylla-specific table metadata.
 schema_ptr scylla_tables(schema_features features) {
-    static auto make = [] (bool has_cdc_options) -> schema_ptr {
+    static auto make = [] (bool has_cdc_options, bool has_per_table_partitioners) -> schema_ptr {
```
I'm sorry, I don't understand anything that's going on in this function, @haaawk I hope you understand what you did here :-) I seriously think at some point we'll need to create a more understandable mechanism for adding features, and document it so it doesn't look like black magic.
That being said, I am saying that I don't understand this - not that I think it's wrong. So I'll work under the assumption that Avi, who also reviewed this code, understood it and it's correct...

haaawk replied:
I'm just extending the mechanism that was previously added for cdc_options so this change is not actually doing much and just relying on already existing mechanism.

avikivity replied:
That's an indication that @nyh is right to worry. A change should be understood, not be such a textual extension. We should encapsulate it into a function with clear explanations.

haaawk replied:
OK. I will as a follow-up PR

avikivity replied:
I didn't mean that you should do it (but you're welcome to do it)

haaawk commented:
@avikivity ping
